### PR TITLE
adding sssom support

### LINKED ISSUE #110: Produce friendlier error from csv_reader on missing column
If a transform's input contains a column not specified in the config, it raises an error through csv_reader:
```
2023-08-03_14:30:57 | WARNING  | koza.io.reader.csv_reader | 'meta'
```
Example: https://github.com/Knowledge-Graph-Hub/kg-phenio/issues/110

This error doesn't state what the issue is but that would be nice to know.
It also seems to appear once for every row in the input (i.e., many times).

### PR DESCRIPTION
**Pull request by glass-ships**
addresses #108 
fix #110 

Tasks:
- [x] Add SSSOM mapping functionality in Koza writers 
- [x] Still need to [do something else](https://github.com/monarch-initiative/koza/pull/109/files#diff-4866988ad23994a9ea0c736fecab24e62d8a24d17a62f6b32d04440f9c647e89R252)
- [x] Add testing (still need narrow match and broad match tests, but in another PR)
- [x] Update documentation 
- [x] ~~Clarify existing examples (this can/should probably be a separate issue)~~ see #113 
- [ ] Add example w/ sssom config

### CODE REVIEW

File: pyproject.toml
Comment (matentzn) [inline pyproject.toml:28]: Why are you pointing to a fork? If you want to work on a branch, we can just give you write access to the main project!

File: src/koza/cli_runner.py
Comment (matentzn) [inline src/koza/cli_runner.py:83]: I would avoid making changes to code that is unrelated to the issue at hand, but no worries.. Just a thought. Makes it easier for me to review the relevant pieces!

File: src/koza/model/config/source_config.py
Comment (matentzn) [inline src/koza/model/config/source_config.py:162]: This should probably be a private method using the leading `_` in the method name, since it depends on a specific internal state of the class?

**Review by matentzn — COMMENTED** [review]
Just lurking feel free to ignore comments

File: pyproject.toml
Comment (matentzn) [reply pyproject.toml:28]: I gave you write access.

File: pyproject.toml
Comment (matentzn) [inline pyproject.toml:27]:
```suggestion
sssom = ">=0.3.40"
```

Some _very_ important changes about the mapping justification IRIs came in in 36 or so.

File: src/koza/model/config/source_config.py
Comment (matentzn) [inline src/koza/model/config/source_config.py:238]: why is this assigned to a local variable?

File: src/koza/model/config/source_config.py
Comment (matentzn) [inline src/koza/model/config/source_config.py:241]: mixed check on local and class variable (len), why?

File: src/koza/model/config/source_config.py
Comment (matentzn) [inline src/koza/model/config/source_config.py:236]: No need to act, but please read up on https://github.com/cthoyt/curies

**Review by matentzn — COMMENTED** [review]
Where does the config live that says: "use narrow matches for node replacement" vs "us only exact matches for node replacement"?

File: src/koza/model/config/source_config.py
Comment (glass-ships) [reply src/koza/model/config/source_config.py:238]: no reason to be honest. same as the mixed check in your other suggestion, i think it just slipped through the crack and was mostly there for my clarity as i was writing it up. 
removed local in favor of class variable

File: src/koza/model/config/source_config.py
Comment (glass-ships) [reply src/koza/model/config/source_config.py:236]: can do! anything in particular i should take a look at?

**Comment by glass-ships** [comment]
> Where does the config live that says: "use narrow matches for node replacement" vs "us only exact matches for node replacement"?

I... don't know if we have one, to be honest

**Comment by glass-ships** [comment]
Oh, so in response to the config question- I think for now we might just bring this in with only "exact match", and open it up to include narrow in another PR

**Comment by matentzn** [comment]
Restricting to exact sounds good as a first safe step, but you may loose quite a bit of relevant Information. I think it's worth making that configurable via some property like "merge_on_property".

File: pyproject.toml
Comment (glass-ships) [reply pyproject.toml]: Unfortunately, upgrading to 0.3.40 breaks the SSSOM test, and I'm not sure why. 
Reverting to 0.3.31 for the time being.

Comment (matentzn) [reply pyproject.toml]: What's the error message? This change is pretty important so you are not hit by surprise in the next pr. most likely the curie map in you SSSOM test has the wrong value for the semapv prefix.

Comment (glass-ships) [reply pyproject.toml]: TLDR:
```
=================================================================================== short test summary info ====================================================================================
FAILED tests/unit/test_sssom_mapping.py::test_basic_mapping - KeyError: "['predicate_modifier'] not in index"
FAILED tests/unit/test_sssom_mapping.py::test_merge_and_filter - KeyError: "['predicate_modifier'] not in index"
FAILED tests/unit/test_sssom_mapping.py::test_exact_match_is_bidirectional - KeyError: "['predicate_modifier'] not in index"
===================================================================== 3 failed, 107 passed, 1 warning in 71.45s (0:01:11) ======================================================================
```

For full output (it's quite long) I would check out [one of the failed github actions](https://github.com/monarch-initiative/koza/actions/runs/6076500179/job/16484614849)

Comment (matentzn) [reply pyproject.toml]: @hrshdhgd can you look into this and advice?

Comment (hrshdhgd) [reply pyproject.toml]: While using `filter_prefixes`, `features` should be provided as name of columns to be grouped by. The default is `KEY_FEATURES = [SUBJECT_ID, PREDICATE_ID, OBJECT_ID, PREDICATE_MODIFIER]` in `sssom`. 

`features` is not provided in the test here and hence the default is assumed. `predicate_modifier` is not a column in the test data and hence the error. simple fix:

In `source_config.py`:
```
new_msdf = filter_prefixes(
            df=new_msdf.df,
            filter_prefixes=filters,
            require_all_prefixes=False,
            features=new_msdf.df.columns
        )
```

I took the liberty of updating the code and updating `sssom` to the latest version. Apologize if it upsets anything.

Comment (matentzn) [reply pyproject.toml]: Thank you @hrshdhgd

**Comment by glass-ships** [comment]
nice one @hrshdhgd thanks for taking a look at that!!

**Comment by glass-ships** [comment]
is there anything preventing this from being merged as is?

**Comment by glass-ships** [comment]
There's a lot of formatting noise in this now, but the relevant changes are in: 

- `model/config/sssom_config.py`
- tsv_writer
- jsonl_writer
- example usage in `tests/unit/test_sssom_mapping.py`

**Review by kevinschaper — APPROVED** [review]

**Pull request merged**
